cache.go

```diff
@@ -0,0 +1,223 @@
+		if len(pth) == 0 {
+			continue
+		}
+		ignoreByPath[pth] = ignore
+	}
+	return ignoreByPath
+}
+
+// expandPath returns every file included in pth (recursively) if it is a dir,
+// if pth is a file it will be returned as an array.
+func expandPath(pth string) ([]string, error) {
+	info, err := os.Lstat(pth)
+	if err != nil {
+		return nil, err
+	}
+
+	var subPaths []string
+	if info.IsDir() {
+		if err := filepath.Walk(pth, func(p string, i os.FileInfo, err error) error {
+			if err != nil {
+				return err
+			}
+			if i.IsDir() {
+				return nil
+			}
+
+			subPaths = append(subPaths, p)
+			return nil
+		}); err != nil {
+			return nil, err
+		}
+	} else {
+		subPaths = append(subPaths, pth)
+	}
+
+	return subPaths, nil
+}
+
+// normalizeIndicatorByPath modifies indicatorByPath:
+// expands both path to cache and indicator path
+// removes the item if any of path to cache or indicator path is not exist or if the indicator is a dir
+// replaces path to cache (if it is a directory) by every file (recursively) in the directory.
+func normalizeIndicatorByPath(indicatorByPath map[string]string) (map[string]string, error) {
+	normalized := map[string]string{}
+	for pth, indicator := range indicatorByPath {
+		if len(indicator) > 0 {
+			var err error
+			indicator, err = pathutil.AbsPath(indicator)
+			if err != nil {
+				return nil, err
+			}
+
+			info, exist, err := pathutil.PathCheckAndInfos(indicator)
+			switch {
+			case err != nil:
+				return nil, err
+			case !exist:
+				log.Warnf("indicator does not exists at: %s", indicator)
+				continue
+			case info.IsDir():
+				log.Warnf("indicator is a directory: %s", indicator)
+				continue
+			}
+		}
+
+		var err error
+		pth, err = pathutil.AbsPath(pth)
+		if err != nil {
+			return nil, err
+		}
+
+		exist, err := pathutil.IsPathExists(pth)
+		if err != nil {
+			return nil, err
+		}
+		if !exist {
+			log.Warnf("path does not exists at: %s", pth)
+			continue
+		}
+
+		subPths, err := expandPath(pth)
+		if err != nil {
+			return nil, err
+		}
+		for _, p := range subPths {
+			normalized[p] = indicator
+		}
+	}
+	return normalized, nil
+}
+
+// normalizeExcludeByPattern modifies excludeByPattern:
+// expands patterns.
+func normalizeExcludeByPattern(excludeByPattern map[string]bool) (map[string]bool, error) {
+	normalized := map[string]bool{}
+	for pattern, exclude := range excludeByPattern {
+		pattern, err := pathutil.AbsPath(pattern)
+		if err != nil {
+			return nil, err
+		}
+
+		normalized[pattern] = exclude
+	}
+	return normalized, nil
+}
+
+// match reports whether the path matches to any of the given ignore items
+// and returns the exclude property of the matching ignore item.
+func match(pth string, excludeByPattern map[string]bool) (bool, bool) {
+	for pattern, exclude := range excludeByPattern {
+		if strings.Contains(pattern, "*") && glob.Glob(pattern, pth) {
+			return true, exclude
+		}
+
+		if !strings.Contains(pattern, "*") && strings.HasPrefix(pth, pattern) {
```

Comment:
is this two ifs mergeable?

Reply:
could be but i fell this way it is a bit more readable